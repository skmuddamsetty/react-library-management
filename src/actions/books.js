import uuid from 'uuid';
// ADD_BOOK
export const addBook = ({
  title = '',
  description = '',
  price = 0,
  publishedAt = 0,
}) => ({
  type: 'ADD_BOOK',
  book: {
    id: uuid(),
    title,
    description,
    price,
    publishedAt,
  },
});
// EDIT_BOOK
export const editBook = (id, updates) => ({
  type: 'EDIT_BOOK',
  id,
  updates,
});
// REMOVE_BOOK
export const removeBook = ({ id } = {}) => ({
  type: 'REMOVE_BOOK',
  id,
});
