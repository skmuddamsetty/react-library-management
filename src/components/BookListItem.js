import React from 'react';
import { Link } from 'react-router-dom';

const BookListItem = ({
  title,
  description,
  price,
  publishedAt,
  id,
  dispatch,
  history,
}) => (
  <div>
    <Link to={`/edit-book/${id}`}>
      <h3>
        {title}
        <span> - {description}</span>
      </h3>
    </Link>
    <p>{price}</p>
    <p>{publishedAt}</p>
  </div>
);

export default BookListItem;
