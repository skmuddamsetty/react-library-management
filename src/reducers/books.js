const booksReducerDefaultState = [];

const booksReducer = (state = booksReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_BOOK':
      return [...state, action.book];
    case 'REMOVE_BOOK':
      let books = [...state];
      books = books.filter((book) => book.id !== action.id);
      return books;
    case 'EDIT_BOOK':
      return state.map((book) => {
        if (book.id === action.id) {
          return { ...book, ...action.updates };
        } else {
          return book;
        }
      });
    case 'SET_BOOKS':
      return action.books;
    default:
      return state;
  }
};

export default booksReducer;
