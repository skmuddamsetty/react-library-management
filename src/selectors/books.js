import moment from 'moment';
// Get Visible Books
const getVisibleBooks = (books, { text, sortBy, startDate, endDate }) => {
  const filteredBooks = books.filter((book) => {
    const publishedAtMoment = moment(book.publishedAt);
    const startDateMatch = startDate
      ? startDate.isSameOrBefore(publishedAtMoment, 'day')
      : true;
    const endDateMatch = endDate
      ? endDate.isSameOrAfter(publishedAtMoment, 'day')
      : true;
    const textMatch = book.title.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  });
  return filteredBooks.sort((a, b) => {
    if (sortBy === 'date') {
      return a.publishedAt < b.publishedAt ? 1 : -1; // decreasing order
    } else if (sortBy === 'price') {
      return a.price < b.price ? 1 : -1;
    }
  });
};

export default getVisibleBooks;
