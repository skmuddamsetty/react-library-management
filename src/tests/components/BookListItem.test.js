import React from 'react';
import { shallow } from 'enzyme';
import BookListItem from '../../components/BookListItem';
import books from '../fixtures/books';

test('should render books list item with book', () => {
  const wrapper = shallow(<BookListItem {...books[0]} />);
  expect(wrapper).toMatchSnapshot();
});
