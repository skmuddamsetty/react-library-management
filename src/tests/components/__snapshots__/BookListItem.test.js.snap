// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render books list item with book 1`] = `
<div>
  <Link
    to="/edit-book/1"
  >
    <h3>
      React Book
      <span>
         - 
        This is the first react book
      </span>
    </h3>
  </Link>
  <p>
    54500
  </p>
  <p>
    0
  </p>
</div>
`;
